internal/unit-info-viewer/skills-progress.lua

```diff
@@ -0,0 +1,194 @@
+        'dwarfmode/ViewSheets/UNIT/Skills/Other',
+
+        'dungeonmode/ViewSheets/UNIT/Skills/Labor',
+        'dungeonmode/ViewSheets/UNIT/Skills/Combat',
+        'dungeonmode/ViewSheets/UNIT/Skills/Social',
+        'dungeonmode/ViewSheets/UNIT/Skills/Other',
+    },
+    frame={w=54, h=20},
+}
+
+function SkillProgressOverlay:init()
+    self.current_uid = -1
+    self.current_unit = nil
+
+    self:addviews{
+        widgets.Label{
+            view_id='annotations',
+            frame={t=0, r=0, w=16},
+            text='',
+            text_pen=COLOR_GRAY,
+        },
+        widgets.ToggleHotkeyLabel{
+            frame={b=0, l=2, w=25},
+            label='Progress Bar',
+            key='CUSTOM_CTRL_B',
+            options={
+                {label='No', value=false, pen=COLOR_WHITE},
+                {label='Yes', value=true, pen=COLOR_YELLOW},
+            },
+            view_id='toggle_progress',
+            initial_option=true
+        },
+        widgets.ToggleHotkeyLabel{
+            frame={b=0, l=28, w=25},
+            label='Experience  ',
+            key='CUSTOM_CTRL_E',
+            options={
+                {label='No', value=false, pen=COLOR_WHITE},
+                {label='Yes', value=true, pen=COLOR_YELLOW},
+            },
+            view_id='toggle_experience',
+            initial_option=true
+        },
+    }
+end
+
+function SkillProgressOverlay:preUpdateLayout()
+    local screen_width, screen_height = dfhack.screen.getWindowSize()
+    self.frame.h = screen_height - 20
+end
+
+function SkillProgressOverlay:onRenderFrame(dc, rect)
+    local annotations = {}
+    if view_sheets.active_id ~= self.current_uid then
+        self.current_uid = view_sheets.active_id
+        self.current_unit = df.unit.find(self.current_uid)
+    end
+    if self.current_unit and self.current_unit.portrait_texpos > 0 then
+        -- If a portrait is present, displace the bars down 2 tiles
+        table.insert(annotations, "\n\n")
+    end
+
+    local progress_bar_needed = not dfhack.world.isAdventureMode() or not dfhack.screen.inGraphicsMode()
+    self.subviews.toggle_progress.disabled = not progress_bar_needed
+    local progress_bar = self.subviews.toggle_progress:getOptionValue() and progress_bar_needed
+    local experience = self.subviews.toggle_experience:getOptionValue()
+
+    local margin = self.subviews.annotations.frame.w
+    local num_elems = self.frame.h // 3 - 1
+    local start = math.min(view_sheets.scroll_position_unit_skill,
+        math.max(0,#view_sheets.unit_skill-num_elems))
+    local max_elem = math.min(#view_sheets.unit_skill-1,
+        view_sheets.scroll_position_unit_skill+num_elems-1)
+    for idx = start, max_elem do
+        local skill = get_skill(idx, self.current_unit)
+        if not skill then
+            table.insert(annotations, "\n\n\n\n")
+            goto continue
+        end
+        local rating = df.skill_rating.attrs[math.max(0, math.min(skill.rating, 19))]
```

Comment:
use enum values here instead of magic constants. e.g. `df.skill_rating.Dabbling` instead of `0`